### PR TITLE
feat(subscriptions): implement generic subscription for create mutations

### LINKED ISSUE #35: Include subscription service for new items added to the CE
Some algorithms that we expect to run within the CE will be run by a partner on all documents that exist in the CE.
As an example, MTG might want to run MFCC calculations on all audio files added to the CE, regardless of who adds these documents.

We are proposing to add a subscription endpoint to receive notifications each time a DigitalDocument is added to the CE. 
We're not sure if the subscription should specify possible filters (e.g. mimetype of "audio/"), or if all subscribed clients should receive all documents and perform filtering locally.

### PR DESCRIPTION
**Pull request by ChristiaanScheermeijer**
## Basic subscriptions

This pull request implements a basic and generic subscription model for node creations in the ce-api.

The following subscription query can be used to listen for DigitalDocuments and MusicComposition creations:

```graphql
subscription {
  ThingCreateMutation(onTypes: [DigitalDocument, MusicComposition]) {
    identifier
    type
  }
}
```

When for example a DigitalDocument gets created (currently only by using the CreateDigitalDocument mutation and not the MergeDigitalDocument mutation) the following response can be expected:

```json
{
  "data": {
    "ThingCreateMutation": {
      "identifier": "0ae38e36-be80-492f-af92-d3212429325c",
      "type": "MusicComposition"
    }
  }
}
```

## MediaObjects

@musicog 

For the following types, I've added specific subscription queries:

- MediaObject
- AudioObject
- VideoObject

All of the above have the option to filter nodes created with a specific `encodingFormat`:

```graphql
subscription {
  MediaObjectCreateMutation(encodingFormat:"video/mp4") {
    identifier
    type
  }
}
```

Currently, if you want to listen for audio and video objects, you'll need two separate subscriptions.

Fixes #35 

## Future

While researching this subject, I've found the following plugins which would solve multiple issues:

https://neo4j.com/labs/kafka/
https://github.com/ancashoria/graphql-kafka-subscriptions

With the two plugins from above, subscriptions will be triggered based on changes in the Neo4j database. This can possibly solve that `Merge` mutations aren't triggering events and advanced filtering can be applied for subscriptions.

### CODE REVIEW

**Comment by ChristiaanScheermeijer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/commands/QueryAndPublishResolver.js  4
         

Clones added
============
- src/resolvers/subscription.js  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/trompamusic/ce-api/pullRequest?prid=4824669&bid=16035544)

**Comment by ChristiaanScheermeijer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/commands/QueryAndPublishResolver.js  4
         

Clones added
============
- src/resolvers/subscription.js  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/trompamusic/ce-api/pullRequest?prid=4824669&bid=16035544)

**Comment by alastair** [comment]
Do you have some client code that you're using to test these subscriptions? We'll write some code for our python library to do the same, but if you already have some code then it will be faster for me to test

**Comment by ChristiaanScheermeijer** [comment]
Sorry @alastair, didn't see your message.

Here are the subscriptions that I've tested in my local environment:

```graphql
subscription {
  MediaObjectCreateMutation {
    identifier
    type
  }
}
```


```graphql
subscription {
  VideoObjectCreateMutation(encodingType:"video/mp4") {
    identifier
    type
  }
}
```


```graphql
subscription {
  AudioObjectCreateMutation(encodingType:"audio/mpeg") {
    identifier
    type
  }
}
```

```graphql
subscription {
  ThingCreateMutation(onTypes: [DigitalDocument, MediaObject]) {
    identifier
    type
  }
}
```

**Comment by ChristiaanScheermeijer** [comment]
@alastair after re-reading your comment, I'm using the GraphiQL interface to test the subscriptions. In other client projects, we are using the default Apollo subscriptions.

There is also the NodeJS algorithm POC which uses the `subscriptions-transport-ws` library: https://github.com/trompamusic/ce-poc-algorithm/blob/master/algorithm/src/index.js#L23

**Comment by ChristiaanScheermeijer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/commands/QueryAndPublishResolver.js  4
         

Clones added
============
- src/resolvers/subscription.js  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/trompamusic/ce-api/pullRequest?prid=4824669&bid=16035544)

**Comment by alastair** [comment]
ah, that's perfect. Thanks.
I've added to my to-do list to test this functionality before your sprint starts.

**Comment by ChristiaanScheermeijer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/resolvers/subscription.js  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/trompamusic/ce-api/pullRequest?prid=4824669&bid=16035544)

File: src/resolvers/mutation.js [hunk 1]
```diff
@@ -13,10 +16,23 @@ export const mutationResolvers = {
...
+    CreateMediaObject: QueryAndPublishResolver.createResolver('MediaObject', 'MediaObjectCreateMutation'),
+    CreateVideoObject: QueryAndPublishResolver.createResolver('VideoObject', 'VideoObjectCreateMutation'),
+    CreateAudioObject: QueryAndPublishResolver.createResolver('AudioObject', 'AudioObjectCreateMutation')
   }
 }
 
+// add resolvers for all create mutations
+walkSync(`${__dirname}/../schema/type`, { directories: false, includeBasePath: true, globs: ['*.graphql'] })
+  .forEach(file => {
+    const { name } = parse(file)
```
Comment (alastair) [inline src/resolvers/mutation.js]: In the case where we have file with multiple type definitions in it (https://github.com/trompamusic/ce-api/issues/93#issuecomment-599540992), it looks like this won't do exactly what we expect.
What's the alternative here? We would have to actually load these files and parse all of the types out of them. Is there a graphql library that allows us to do this?

Comment (ChristiaanScheermeijer) [reply src/resolvers/mutation.js]: Good point @alastair, I could use the same `graphql-tools` function to iterate all types.

https://github.com/trompamusic/ce-api/pull/98/files#diff-1b21ae31202733dc0c9c48a31aa51040R19-R33

**Comment by ChristiaanScheermeijer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/resolvers/subscription.js  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/trompamusic/ce-api/pullRequest?prid=4824669&bid=16035544)

**Comment by ChristiaanScheermeijer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
           

Clones added
============
- src/resolvers/subscription.js  5
- src/transformers/authenticationFieldTransformer.js  1
- src/transformers/subscriptionFieldTransformer.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/trompamusic/ce-api/pullRequest?prid=4824669&bid=16035544)

File: src/resolvers/mutation.js (hunk 1 above)
Comment (ChristiaanScheermeijer) [reply src/resolvers/mutation.js]: @alastair this has been fixed in the latest commit https://github.com/trompamusic/ce-api/pull/51/commits/e0fc5265df8ba7874f9c14d4f8dee815cbaab239

Comment (alastair) [reply src/resolvers/mutation.js]: great. I'm updating our python library to support these subscriptions today to test the rest of the functionality

Comment (ChristiaanScheermeijer) [reply src/resolvers/mutation.js]: Cool, I will merge this feature so we can get things going on our end. I will create a separate PR for any feedback you might have.

Thanks!

**Pull request merged**